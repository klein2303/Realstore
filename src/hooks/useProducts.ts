import {useQuery, UseQueryResult} from "@tanstack/react-query";
import {Product} from "../models/Product";

const fetchProducts = async (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): Promise<Product[]> => {
    let url = "https://fakestoreapi.com/products";

    if (productId) {
        url = `${url}/${productId}`;
        const response = await fetch(url);
        if (!response.ok) throw new Error("Error fetching product");
        const product: Product = await response.json();
        return [product]; // Return a single product wrapped in an array
    } else if (category) {
        url = `${url}/category/${category}`;
    }

    const response = await fetch(url);
    if (!response.ok) throw new Error("Error fetching products");
    let products: Product[] = await response.json();

    // Filter products based on price and rating
    if (minPrice !== undefined) {
        products = products.filter((product) => product.price >= minPrice);
    }
    if (maxPrice !== undefined) {
        products = products.filter((product) => product.price <= maxPrice);
    }
    if (minRating !== undefined) {
        products = products.filter((product) => product.rating.rate >= minRating);
    }

    return products;
};

// Custom hook to fetch products with optional filters
export const useProducts = (
    category?: string,
    productId?: number,
    minPrice?: number,
    maxPrice?: number,
    minRating?: number,
): UseQueryResult<Product[]> => {
    const queryKey = ["products", category, productId, minPrice, maxPrice, minRating];

    return useQuery<Product[]>({
        queryKey,
        queryFn: () => fetchProducts(category, productId, minPrice, maxPrice, minRating),
    });
};
