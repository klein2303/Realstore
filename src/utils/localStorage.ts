// Function to store favorite items in local storage
export const setFavorites = (items: number[]): void => {
    localStorage.setItem("favorites", JSON.stringify(items));
};

// Function to get favorite items from local storage
export const getFavorites = (): number[] => {
    const storedItems = localStorage.getItem("favorites");
    return storedItems ? JSON.parse(storedItems) : [];
};

export const toggleFavorite = (productId: number): void => {
    const favorites = getFavorites();
    const isFavorite = favorites.includes(productId);

    const updatedFavorites = isFavorite ? favorites.filter((id) => id !== productId) : [...favorites, productId];

    setFavorites(updatedFavorites);
};
