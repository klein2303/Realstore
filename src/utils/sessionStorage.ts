import {Filter} from "../pages/Shopping";

export type CartItem = {productId: number; quantity: number};

// Helper function to store items in session storage
const setSessionStorageItems = <T>(key: string, items: T[]): void => {
    sessionStorage.setItem(key, JSON.stringify(items));
};

// Helper function to get items from session storage
const getSessionStorageItems = <T>(key: string): T[] => {
    const storedItems = sessionStorage.getItem(key);
    return storedItems ? JSON.parse(storedItems) : [];
};

// Cart-specific functions
export const getCart = (): CartItem[] => {
    return getSessionStorageItems<CartItem>("cart");
};

export const addToCart = (productId: number, quantity: number = 1): void => {
    const cart = getCart();
    const existingItem = cart.find((item) => item.productId === productId);

    if (existingItem) {
        existingItem.quantity += quantity; // Increment quantity
    } else {
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart);
};

export const clearCart = (): void => {
    setSessionStorageItems("cart", []);
};

export const updateCartQuantity = (productId: number, quantity: number): void => {
    const cart = getCart();
    const index = cart.findIndex((item) => item.productId === productId);

    if (index !== -1) {
        // Product exists in the cart
        if (quantity > 0) {
            cart[index].quantity = quantity; // Update quantity
        } else {
            cart.splice(index, 1); // Remove if quantity is 0
        }
    } else if (quantity > 0) {
        // Product doesn't exist in the cart, add it
        cart.push({productId, quantity});
    }

    setSessionStorageItems("cart", cart); // Save updated cart to sessionStorage
};

// Filtered items-specific functions
export const getFilter = (): Filter[] => {
    return getSessionStorageItems<Filter>("filter");
};

export const setFilter = (filter: Filter[]): void => {
    setSessionStorageItems("filter", filter);
};

export const clearFilter = (): void => {
    setSessionStorageItems("filter", []);
};
