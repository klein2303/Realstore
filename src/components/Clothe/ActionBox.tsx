import {PiHeartStraightThin, PiHeartStraightFill} from "react-icons/pi";
import styles from "./ActionBox.module.css";
import {Link} from "react-router-dom";
import {useState} from "react";
import {toggleFavorite, getFavorites} from "../../utils/localStorage";
import {addToCart, CartItem, getCart} from "../../utils/sessionStorage";
import {useProducts} from "../../hooks/useProducts";

interface ActionBoxProps {
    productId: number;
}

const ActionBox = ({productId}: ActionBoxProps) => {
    //Ensures only one item is added to cart
    const quantity: number = 1;

    const [isFavorited, setIsFavorited] = useState<boolean>(false);
    const [isAddedToCart, setIsAddedToCart] = useState<boolean>(false);

    //Get the specific item with the hook
    const {data: products, isLoading, isError} = useProducts(undefined, productId, undefined, undefined, undefined);

    if (isLoading) {
        return (
            <section role="status" aria-live="polite">
                Loading...
            </section>
        );
    }
    if (isError) {
        return (
            <section role="alert" aria-live="assertive">
                Error fetching products.
            </section>
        );
    }

    const product = Array.isArray(products) ? products.find((p) => p.id === productId) : null;

    const favorites = getFavorites();
    const isAlreadyFavorited = favorites.includes(productId);

    const productsInCart = getCart();
    const isAlreadyInCart = productsInCart.some((product: CartItem) => product.productId === productId);

    if (!isAddedToCart && isAlreadyInCart) {
        setIsAddedToCart(true);
    }

    const addFavorite = (productId: number) => {
        if (!isAlreadyFavorited) {
            toggleFavorite(productId);
            console.log("Added this product as favorite:", product?.title);
        } else {
            toggleFavorite(productId);
            console.log("Removed this product as favorite:", product?.title);
        }
        setIsFavorited(!isFavorited);
    };

    const addProductToCart = (productId: number, quantity: number) => {
        addToCart(productId, quantity);
        setIsAddedToCart(true);
    };

    return (
        <>
            <nav className={styles.link} aria-label="Navigation">
                <Link to={"/"} className={styles.linkstyle}>
                    Home
                </Link>{" "}
                /
                <Link to={"/shopping"} className={styles.linkstyle}>
                    {" "}
                    Shop
                </Link>{" "}
                /
                <p className={styles.pstyle} id="product-title">
                    {" "}
                    {product?.title?.split(" ").slice(0, 3).join(" ")}
                </p>
            </nav>
            <h1 className={styles.title}>SHOP - {product?.title}</h1>
            <section className={styles.content} role="region" aria-label="product-information">
                <img src={product?.image} alt={product?.title} className={styles.imageView} />
                <article className={styles.information} role="region">
                    <section className={styles.productInfoAndHeart}>
                        <section className={styles.productInfo}>
                            <p className={styles.titleInDesc} id="product-description">
                                {product?.title}
                            </p>
                            <p className={styles.price}>{product?.price} $</p>
                            <p className={styles.taxes}>Taxes are included</p>
                        </section>
                        {isAlreadyFavorited ? (
                            <PiHeartStraightFill
                                className={`${styles.heartButton} ${styles.likedHeartButton}`}
                                onClick={() => addFavorite(productId)}
                                tabIndex={0}
                                onKeyDown={(e) => e.key === "Enter" && addFavorite(productId)}
                                aria-label="Remove from favorites"
                                role="button"
                            />
                        ) : (
                            <PiHeartStraightThin
                                className={`${styles.heartButton} ${styles.notLikedheartButton}`}
                                onClick={() => addFavorite(productId)}
                                tabIndex={0}
                                onKeyDown={(e) => e.key === "Enter" && addFavorite(productId)}
                                aria-label="Add to favorites"
                                role="button"
                            />
                        )}
                    </section>
                    <p className={styles.clotheDesc}>{product?.description}</p>
                    {isAddedToCart ? (
                        <div className={styles.alreadyInCart} role="status" aria-live="polite">
                            ADDED TO CART
                        </div>
                    ) : (
                        <button
                            className={styles.cartButton}
                            onClick={() => addProductToCart(productId, quantity)}
                            aria-label="Add to cart">
                            ADD TO CART
                        </button>
                    )}
                </article>
            </section>
        </>
    );
};

export default ActionBox;
