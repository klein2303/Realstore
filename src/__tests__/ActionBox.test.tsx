import {render, screen, fireEvent} from "@testing-library/react";
import {describe, it, expect, beforeEach, vi} from "vitest";
import ActionBox from "../components/Clothe/ActionBox";
import {useProducts} from "../hooks/useProducts";
import {getFavorites} from "../utils/localStorage";
import {getCart, addToCart} from "../utils/sessionStorage";
import {MemoryRouter} from "react-router-dom";

const setScreenSize = (width: number, height: number) => {
    window.innerWidth = width;
    window.innerHeight = height;
    window.dispatchEvent(new Event("resize"));
};

// Mock the hooks and utility functions
vi.mock("../hooks/useProducts", () => ({
    useProducts: vi.fn(),
}));

vi.mock("../utils/localStorage", () => ({
    getFavorites: vi.fn(),
    toggleFavorite: vi.fn(),
}));

vi.mock("../utils/sessionStorage", () => ({
    getCart: vi.fn(),
    addToCart: vi.fn(),
}));

beforeEach(() => {
    vi.resetAllMocks();

    (getFavorites as jest.Mock).mockReturnValue([]);
    (getCart as jest.Mock).mockReturnValue([]);
});

describe("ActionBox component", () => {
    it("renders loading state initially", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: true,
            isError: false,
        });

        const {container} = render(
            <MemoryRouter>
                <ActionBox productId={1} />
            </MemoryRouter>,
        );

        expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
        expect(container).toMatchSnapshot(); // Snapshot test
    });

    it("renders error state when product fetching fails", () => {
        (useProducts as jest.Mock).mockReturnValue({
            data: null,
            isLoading: false,
            isError: true,
        });

        const {container} = render(
            <MemoryRouter>
                <ActionBox productId={1} />
            </MemoryRouter>,
        );

        expect(screen.getByText(/Error fetching products./i)).toBeInTheDocument();
        expect(container).toMatchSnapshot(); // Snapshot test
    });

    it("renders product details and allows adding to cart", () => {
        const fakeProducts = [
            {
                id: 1,
                title: "Test Product",
                price: 99.99,
                description: "Test description",
                image: "test-image.jpg",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: fakeProducts,
            isLoading: false,
            isError: false,
        });

        (getFavorites as jest.Mock).mockReturnValue([]);
        (getCart as jest.Mock).mockReturnValue([]);

        render(
            <MemoryRouter>
                <ActionBox productId={1} />
            </MemoryRouter>,
        );

        // More specific queries for elements
        const productTitle = screen.getByRole("heading", {name: /Test Product/i});
        expect(productTitle).toBeInTheDocument();

        const productPrice = screen.getByText(/99.99 \$/i);
        expect(productPrice).toBeInTheDocument();

        const productImage = screen.getByAltText(/Test Product/i);
        expect(productImage).toHaveAttribute("src", "test-image.jpg");

        const addToCartButton = screen.getByRole("button", {name: /add to cart/i});
        fireEvent.click(addToCartButton);

        expect(addToCart).toHaveBeenCalledWith(1, 1);
    });

    it("renders correctly on mobile version", () => {
        setScreenSize(375, 667); // iPhone 6/7/8 size

        const fakeProducts = [
            {
                id: 1,
                title: "Sample Product",
                price: "$10.00",
                description: "Sample Description",
                image: "https://via.placeholder.com/150",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: fakeProducts,
            isLoading: false,
            isError: false,
        });

        const {container} = render(
            <MemoryRouter>
                <ActionBox productId={1} />
            </MemoryRouter>,
        );

        const elements = screen.getAllByAltText("Sample Product");
        expect(elements[0]).toBeVisible();
        expect(container).toMatchSnapshot(); // Snapshot test
    });

    it("renders correctly on desktop version", () => {
        setScreenSize(1920, 1080); // Desktop size

        const fakeProducts = [
            {
                id: 1,
                title: "Sample Product",
                price: "$10.00",
                description: "Sample Description",
                image: "https://via.placeholder.com/150",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: fakeProducts,
            isLoading: false,
            isError: false,
        });

        const {container} = render(
            <MemoryRouter>
                <ActionBox productId={1} />
            </MemoryRouter>,
        );

        const elements = screen.getAllByAltText("Sample Product");
        expect(elements[0]).toBeVisible();
        expect(container).toMatchSnapshot(); // Snapshot test
    });

    it('adds product to cart when "Add to Cart" button is clicked, for desktop version', () => {
        const fakeProducts = [
            {
                id: 1,
                title: "Sample Product",
                price: "$10.00",
                description: "Sample Description",
                image: "https://via.placeholder.com/150",
            },
        ];

        (useProducts as jest.Mock).mockReturnValue({
            data: fakeProducts,
            isLoading: false,
            isError: false,
        });

        const {container} = render(
            <MemoryRouter>
                <ActionBox productId={1} />
            </MemoryRouter>,
        );

        setScreenSize(1920, 1080); // Set screen size for desktop

        const addToCartButton = screen.getByRole("button", {name: /add to cart/i});
        fireEvent.click(addToCartButton);

        expect(addToCart).toHaveBeenCalledWith(1, 1);
        expect(container).toMatchSnapshot(); // Snapshot test
    });
});
