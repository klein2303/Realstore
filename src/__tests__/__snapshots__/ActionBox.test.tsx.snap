// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionBox component > adds product to cart when "Add to Cart" button is clicked, for desktop version 1`] = `
<div>
  <nav
    aria-label="Navigation"
    class="_link_1ba521"
  >
    <a
      class="_linkstyle_1ba521"
      href="/"
    >
      Home
    </a>
     
    /
    <a
      class="_linkstyle_1ba521"
      href="/shopping"
    >
       
      Shop
    </a>
     
    /
    <p
      class="_pstyle_1ba521"
      id="product-title"
    >
       
      Sample Product
    </p>
  </nav>
  <h1
    class="_title_1ba521"
  >
    SHOP - 
    Sample Product
  </h1>
  <section
    aria-label="product-information"
    class="_content_1ba521"
    role="region"
  >
    <img
      alt="Sample Product"
      class="_imageView_1ba521"
      src="https://via.placeholder.com/150"
    />
    <article
      class="_information_1ba521"
      role="region"
    >
      <section
        class="_productInfoAndHeart_1ba521"
      >
        <section
          class="_productInfo_1ba521"
        >
          <p
            class="_titleInDesc_1ba521"
            id="product-description"
          >
            Sample Product
          </p>
          <p
            class="_price_1ba521"
          >
            $10.00
             $
          </p>
          <p
            class="_taxes_1ba521"
          >
            Taxes are included
          </p>
        </section>
        <svg
          aria-label="Add to favorites"
          class="_heartButton_1ba521 _notLikedheartButton_1ba521"
          fill="currentColor"
          height="1em"
          role="button"
          stroke="currentColor"
          stroke-width="0"
          tabindex="0"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M220.18,59.79a54.05,54.05,0,0,0-76.31,0L128,74.51,112.13,59.74A54,54,0,0,0,35.8,136.15l89.35,90.66a4,4,0,0,0,5.7,0l89.33-90.64a54,54,0,0,0,0-76.38Zm-5.67,70.74L128,218.3,41.47,130.51a46,46,0,0,1,65.06-65.06l.1.1,18.64,17.36a4,4,0,0,0,5.46,0l18.64-17.36.1-.1a46,46,0,1,1,65,65.08Z"
          />
        </svg>
      </section>
      <p
        class="_clotheDesc_1ba521"
      >
        Sample Description
      </p>
      <div
        aria-live="polite"
        class="_alreadyInCart_1ba521"
        role="status"
      >
        ADDED TO CART
      </div>
    </article>
  </section>
</div>
`;

exports[`ActionBox component > renders correctly on desktop version 1`] = `
<div>
  <nav
    aria-label="Navigation"
    class="_link_1ba521"
  >
    <a
      class="_linkstyle_1ba521"
      href="/"
    >
      Home
    </a>
     
    /
    <a
      class="_linkstyle_1ba521"
      href="/shopping"
    >
       
      Shop
    </a>
     
    /
    <p
      class="_pstyle_1ba521"
      id="product-title"
    >
       
      Sample Product
    </p>
  </nav>
  <h1
    class="_title_1ba521"
  >
    SHOP - 
    Sample Product
  </h1>
  <section
    aria-label="product-information"
    class="_content_1ba521"
    role="region"
  >
    <img
      alt="Sample Product"
      class="_imageView_1ba521"
      src="https://via.placeholder.com/150"
    />
    <article
      class="_information_1ba521"
      role="region"
    >
      <section
        class="_productInfoAndHeart_1ba521"
      >
        <section
          class="_productInfo_1ba521"
        >
          <p
            class="_titleInDesc_1ba521"
            id="product-description"
          >
            Sample Product
          </p>
          <p
            class="_price_1ba521"
          >
            $10.00
             $
          </p>
          <p
            class="_taxes_1ba521"
          >
            Taxes are included
          </p>
        </section>
        <svg
          aria-label="Add to favorites"
          class="_heartButton_1ba521 _notLikedheartButton_1ba521"
          fill="currentColor"
          height="1em"
          role="button"
          stroke="currentColor"
          stroke-width="0"
          tabindex="0"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M220.18,59.79a54.05,54.05,0,0,0-76.31,0L128,74.51,112.13,59.74A54,54,0,0,0,35.8,136.15l89.35,90.66a4,4,0,0,0,5.7,0l89.33-90.64a54,54,0,0,0,0-76.38Zm-5.67,70.74L128,218.3,41.47,130.51a46,46,0,0,1,65.06-65.06l.1.1,18.64,17.36a4,4,0,0,0,5.46,0l18.64-17.36.1-.1a46,46,0,1,1,65,65.08Z"
          />
        </svg>
      </section>
      <p
        class="_clotheDesc_1ba521"
      >
        Sample Description
      </p>
      <button
        aria-label="Add to cart"
        class="_cartButton_1ba521"
      >
        ADD TO CART
      </button>
    </article>
  </section>
</div>
`;

exports[`ActionBox component > renders correctly on mobile version 1`] = `
<div>
  <nav
    aria-label="Navigation"
    class="_link_1ba521"
  >
    <a
      class="_linkstyle_1ba521"
      href="/"
    >
      Home
    </a>
     
    /
    <a
      class="_linkstyle_1ba521"
      href="/shopping"
    >
       
      Shop
    </a>
     
    /
    <p
      class="_pstyle_1ba521"
      id="product-title"
    >
       
      Sample Product
    </p>
  </nav>
  <h1
    class="_title_1ba521"
  >
    SHOP - 
    Sample Product
  </h1>
  <section
    aria-label="product-information"
    class="_content_1ba521"
    role="region"
  >
    <img
      alt="Sample Product"
      class="_imageView_1ba521"
      src="https://via.placeholder.com/150"
    />
    <article
      class="_information_1ba521"
      role="region"
    >
      <section
        class="_productInfoAndHeart_1ba521"
      >
        <section
          class="_productInfo_1ba521"
        >
          <p
            class="_titleInDesc_1ba521"
            id="product-description"
          >
            Sample Product
          </p>
          <p
            class="_price_1ba521"
          >
            $10.00
             $
          </p>
          <p
            class="_taxes_1ba521"
          >
            Taxes are included
          </p>
        </section>
        <svg
          aria-label="Add to favorites"
          class="_heartButton_1ba521 _notLikedheartButton_1ba521"
          fill="currentColor"
          height="1em"
          role="button"
          stroke="currentColor"
          stroke-width="0"
          tabindex="0"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M220.18,59.79a54.05,54.05,0,0,0-76.31,0L128,74.51,112.13,59.74A54,54,0,0,0,35.8,136.15l89.35,90.66a4,4,0,0,0,5.7,0l89.33-90.64a54,54,0,0,0,0-76.38Zm-5.67,70.74L128,218.3,41.47,130.51a46,46,0,0,1,65.06-65.06l.1.1,18.64,17.36a4,4,0,0,0,5.46,0l18.64-17.36.1-.1a46,46,0,1,1,65,65.08Z"
          />
        </svg>
      </section>
      <p
        class="_clotheDesc_1ba521"
      >
        Sample Description
      </p>
      <button
        aria-label="Add to cart"
        class="_cartButton_1ba521"
      >
        ADD TO CART
      </button>
    </article>
  </section>
</div>
`;

exports[`ActionBox component > renders error state when product fetching fails 1`] = `
<div>
  <section
    aria-live="assertive"
    role="alert"
  >
    Error fetching products.
  </section>
</div>
`;

exports[`ActionBox component > renders loading state initially 1`] = `
<div>
  <section
    aria-live="polite"
    role="status"
  >
    Loading...
  </section>
</div>
`;
